h2o-algos/src/main/java/hex/glm/ComputationState.java

```diff
@@ -265,25 +328,78 @@ public GLMSubsetGinfo(GLMGradientInfo fullInfo, int N, int c, int [] ids) {
       super(fullInfo._likelihood, fullInfo._objVal, extractSubRange(N,c,ids,fullInfo._gradient));
       _fullInfo = fullInfo;
     }
+
+    public GLMSubsetGinfo(GLMGradientInfo fullInfo) {
+      super(fullInfo._likelihood, fullInfo._objVal, fullInfo._gradient);
+      _fullInfo = fullInfo;
+    }
+
+    public GLMSubsetGinfo(GLMGradientInfo fullInfo, double[] gradient) {
+      super(fullInfo._likelihood, fullInfo._objVal, gradient);
+      _fullInfo = fullInfo;
+    }
   }
   public GradientSolver gslvrMultinomial(final int c) {
     final double [] fullbeta = _beta.clone();
     return new GradientSolver() {
       @Override
-      public GradientInfo getGradient(double[] beta) {
-        fillSubRange(_activeData.fullN()+1,c,_activeDataMultinomial[c].activeCols(),beta,fullbeta);
-        GLMGradientInfo fullGinfo =  _gslvr.getGradient(fullbeta);
-        return new GLMSubsetGinfo(fullGinfo,_activeData.fullN()+1,c,_activeDataMultinomial[c].activeCols());
+      public GradientInfo getGradient(double[] beta) { // beta stores coeff of one class for other
+        if (_multinomialSpeedup) {  // beta only contains active coefficients of all classes, need to expand it back to fullbeta
```

Comment:
I would very much extract the IF outside of the class and used inharitance to separate the two if branches